Implement a message queue between gateway and workflow service for improved reliability

rikukissa opened the pull request:
## To-do

- [x] Implement a new endpoint for country config to reject a registration (like confirm registration). The current synchronous flow where the `/event-registration` can return anything other than `2**` status to indicate the request got accepted needs to be refactored. Endpoint needs to confirm the message was received and then call another endpoint in core to let core know if the record got rejected or accepted.
  - [x] Done here: https://github.com/opencrvs/opencrvs-core/pull/7674 
- [x] Allow configuring a port, username and password for Redis connection. Let services pass a full `RedisOptions` object inside `useRecordQueue`

- [ ] If the services are restarted, the token expires in queue the same issue can happen if a message is in the queue for too long and the user token expires. Instead of us storing the token in the message queue payloads, let's instead store the initiating user's id and roll a new token in the code the message queue triggers. 

- [ ] Fix conflicts on this branch

- [ ] **Investigate the failing tests:** should they be removed or rewritten?
  - [ ] Fix the tests
 
## Dependencies
- https://github.com/opencrvs/opencrvs-core/issues/6279
  - This needs to be done so that we can implement an endpoint to the auth service to request a token for a given user id without the endpoint being exposed to public internet like we have all of them today.

## Technical tests
- [ ] How does Redis queue fair with large attachments?
- [ ] How does both the queue and frontend behave when you create multiple drafts offline and then come back online

## Tips

Run this to see a local GUI for your message queue items

```
docker run -p 3003:3000 -e REDIS_HOST=host.docker.internal  deadly0/bull-board
```

rikukissa committed:
implement idempotent, transactional writes for services, add a message queue between gateway and workflow service

rikukissa commented on packages/commons/src/fhir/task.ts:
```diff
@@ -111,10 +114,14 @@ export type Task = Omit<
   lastModified: string
   status: 'ready' | 'requested' | 'draft' | 'accepted' | 'rejected'
   extension: Array<Extension>
+  focus: {
```
Just a random type improvement for Task as it came up as part of the task. Don't think there is ever a case where a Task wouldn't reference a composition in `focus.reference`

rikukissa committed:
Merge branch 'develop' of github.com:opencrvs/opencrvs-core into ocrvs-6208

rikukissa commented on packages/commons/src/message-queue/index.ts:
```diff
@@ -0,0 +1,12 @@
+ * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
+ */
+
+export * from './record'
```
This is our new message queue module :) All posting to and listening from the queues should happen through this module in commons. This ensures that the type chain doesn't break in the code.

rikukissa commented on packages/gateway/src/workflow/index.ts:
```diff
@@ -129,13 +135,32 @@ export async function createRegistration(
-    trackingId: string
-    isPotentiallyDuplicate: boolean
-  }>('POST', '/create-record', authHeader, { record, event })
+  if (inScope(authHeader, [USER_SCOPE.REGISTER])) {
```
Remember how the "create record" handler is basically 3 operations in one? I split it and lifted the detection to this level. Ideally the client would call right graphql handler from the start

rikukissa commented on packages/metrics/package.json:
```diff
@@ -25,6 +25,7 @@
     "csv-stringify": "^5.3.4",
     "csv-writer": "^1.6.0",
     "date-fns": "^2.28.0",
+    "envalid": "^8.0.0",
```
This was added, rest are just reordered

rikukissa commented on packages/metrics/src/api.ts:
```diff
@@ -21,16 +21,14 @@ import { logger } from '@opencrvs/commons'
     headers: {
-      'Content-Type': 'application/fhir+json',
-      ...authHeader
+      'Content-Type': 'application/fhir+json'
```
Ok, here starts some not-very-major but many-in-numbers changes start. I removed the way we send auth tokens to Hearth as it doesn't use it. for anything. More of a risk than a benefit if nothing in the receiving end uses it

rikukissa replied:
Only token removals in this file.

rikukissa commented on packages/metrics/src/features/audit/handler.ts:
```diff
@@ -27,9 +27,14 @@ export async function newAuditHandler(
+
     const payload = request.payload as IUserAuditBody
+
+    const transactionId = payload.transactionId
```
Transaction IDs are used to determine if an operation has already been performed. A service can for instance check if there is already a database row with a certain ID. If yes, then it wont write it again

rikukissa commented on packages/metrics/src/features/audit/handler.ts:
```diff
@@ -50,7 +55,8 @@ export async function newAuditHandler(
     )
-    await writePoints(points)
+
+    await writePoints(points, transactionId)
```
`writePoints` function is now in charge of never writing duplicates even if the handler is called multiple times (given transaction id stays the same)

rikukissa commented on packages/metrics/src/features/export/handler.ts:
```diff
@@ -61,9 +61,7 @@ export async function monthlyExportHandler(
   if (monthlyMetrics.genderBasisMetrics) {
```
Only token removals in this file

rikukissa commented on packages/metrics/src/features/locationStatistics/service.ts:
```diff
@@ -9,15 +9,15 @@
-  fetchChildLocationsByParentId,
   countRegistrarsByLocation,
-  totalOfficesInCountry,
+  fetchChildLocationsByParentId,
```
Only token removals in this file

rikukissa commented on packages/metrics/src/features/locationWiseEventEstimations/handler.ts:
```diff
@@ -44,8 +44,7 @@ export async function locationWiseEventEstimationsHandler(
   const childLocations = await fetchChildLocationsByParentId(
-    locationId || 'Location/0',
-    authHeader
+    locationId || 'Location/0'
```
Only token removals in this file

rikukissa commented on packages/metrics/src/features/metrics/handler.ts:
```diff
@@ -71,8 +71,7 @@ export async function metricsHandler(
     currentLocationLevel,
-    lowerLocationLevel,
-    authHeader
+    lowerLocationLevel
```
Only token removals in this file

rikukissa commented on packages/metrics/src/features/metrics/utils.test.ts:
```diff
@@ -243,8 +243,7 @@ describe('verify metrics util', () => {
       const result = await getDistrictLocation(
-        'b2b3ca8b-a14f-41c6-b97f-7cb99a1299e5',
-        { Authorization: 'bearer token' }
+        'b2b3ca8b-a14f-41c6-b97f-7cb99a1299e5'
```
Only token removals in this file

rikukissa commented on packages/metrics/src/features/metrics/utils.ts:
```diff
@@ -268,7 +268,7 @@ export const fetchEstimateForTargetDaysByLocationId = async (
 ): Promise<IEstimation> => {
   if (locationId) {
-    const locationData: Location = await fetchFHIR(locationId, authHeader)
+    const locationData: Location = await fetchFHIR(locationId)
```
Only token removals in this file

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExeHVjbm1wM3prc3dkZm01a2JtdnA4eTV4ZjRsNHBqcXB3NnhnbmIzZyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9dg/BGVAiLhU93r0uAl4Qj/giphy.gif)

rikukissa committed:
cleanup

rikukissa commented on packages/workflow/src/config/routes.ts:
```diff
@@ -69,15 +68,6 @@ export const getRoutes = () => {
           'Register event based on tracking id and registration number.'
       }
     },
-    {
```
No need for this anymore as the trigger comes from BullMQ

rikukissa commented on packages/workflow/src/records/handler/create.ts:
```diff
@@ -8,39 +8,50 @@
  * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
  */
-import * as Hapi from '@hapi/hapi'
+
```
These changes are the most significant in this PR so please read **extremely carefully**, compare to the old implementation and make parts of the flows fail intentionally by adding for instance

```
if(Math.random() > 0.00005) {
  throw new Error('Random error!')
}
```

To see current status of BullMQ, open up the dashboard by running

```
docker run -p 3003:3000 -e BULLMQ_DASHBOARD=true taskforcesh/bullmq-dashboard
```

and then opening up http://localhost:3003

rikukissa commented on packages/workflow/src/records/handler/create.ts:
```diff
@@ -222,149 +240,385 @@ async function createRecord(
-export default async function createRecordHandler(
-  request: Hapi.Request,
-  _: Hapi.ResponseToolkit
+export async function declareRecordHandler(
```
This function and the two other handlers `validateRecordHandler` and `registerRecordHandler` follow one simple rule: you can call these functions as many times you like with the same parameters and the end result is always the same. So call this function once or million times, you should always only see one record in the database. Idempotency like this is absolutely critical so that the message queue can submit the same record here twice.

Zangetsu101 replied:
I mean are the notification endpoints idempotent too?

rikukissa replied:
Not yet. I left them untouched for now as in many cases it's the last operation to be performed. We should make them idempotent though. I'll add it to the TODO list

rikukissa commented on packages/workflow/src/utils/auth-utils.ts:
```diff
@@ -10,16 +10,7 @@
-  RECORD_SEARCH = 'recordsearch',
-  VERIFY = 'verify'
-}
+import { PlainToken } from '@opencrvs/commons'
```
TokenWithBearer = explicit type for `Bearer abc`
PlainToken = explicit type for  `abc`

rikukissa committed:
fix tests

rikukissa commented on packages/search/src/features/registration/death/handler.test.ts:
```diff
@@ -43,25 +43,6 @@ describe('Verify handlers', () => {
       mockedsearchForDuplicates.mockReturnValue([])
     })
 
-    it('should return status code 500 if invalid payload received', async () => {
```
These were causing stderr output in test runner, so I just removed the test. There shouldn't be any expected case where the server returns a 500 status

rikukissa committed:
cleanup

rikukissa committed:
implement client-side handling for not receiving an id right away for record submission

rikukissa committed:
Merge branch 'develop' of github.com:opencrvs/opencrvs-core into ocrvs-6208

rikukissa committed:
remove unused redis host env variable from search and metrics

rikukissa committed:
make MOSIP_TOKEN_SEEDER_URL optional variable

Zangetsu101 reviewed (commented):
Took me a while to get my head around the record-queue. But awesome stuff Riku!

Zangetsu101 commented on packages/metrics/src/features/registration/handler.ts:
```diff
@@ -74,164 +78,135 @@ export async function markedAsDuplicate(
+    ? Events.WAITING_EXTERNAL_VALIDATION
+    : Events.READY_FOR_REVIEW
 
-  await createUserAuditPointFromFHIR('DECLARED', request)
```
I'm guessing all the "create-record" related ones were separated out?

rikukissa replied:
@Zangetsu101 yea, I only changed the ones that were part of the create record flows. But in the long run, all endpoints should either create a record audit log item OR a user audit log item but never both

rikukissa committed:
Merge branch 'develop' into ocrvs-6208

rikukissa committed:
use more explicit naming for queue events

rikukissa committed:
Merge branch 'ocrvs-6208' of github.com:opencrvs/opencrvs-core into ocrvs-6208

rikukissa commented on packages/workflow/src/utils/duplicate-checker.ts:
```diff
@@ -68,85 +65,112 @@ type BirthDuplicateSearchBody = {
   return []
 }
 
+export function hasSameDuplicatesInExtension(
+  task: SavedTask,
+  duplicateIds: { id: string; trackingId: string }[]
+) {
+  return task.extension?.some(
+    (ext) =>
+      ext.url === FLAGGED_AS_POTENTIAL_DUPLICATE &&
+      ext.valueString ===
+        duplicateIds.map((duplicate) => duplicate.trackingId).toString()
+  )
+}
+
```
No, I don't think so. That's the assumption I built this with

rikukissa commented on packages/workflow/src/records/handler/create.ts:
```diff
@@ -222,149 +240,385 @@ async function createRecord(
+
+  await createUserAuditEvent('REGISTERED', {
+    transactionId: transactionId,
     compositionId: getComposition(record).id,
     trackingId: getTrackingIdFromRecord(record),
-    isPotentiallyDuplicate: false
-  }
+    headers: getAuthorizationHeaderFromToken(token)
+  })
```
@Zangetsu101 you are onto something. Or should it be nothing at all here? I was trying to find `WAITING_EXTERNAL_VALIDATION` from user audit events from the old implementation but couldn't. 

Also a little bit of a conceptual question – does the user action initiation for "registered" happen already before the external validation?

Zangetsu101 replied:
True, if we think of it like that we _are_ tracking the user actions and the user _is_ initiating the "register" aciton

rikukissa commented on packages/workflow/src/records/handler/create.ts:
```diff
@@ -222,149 +240,385 @@ async function createRecord(
+
+  const existingComposition = draftId
+    ? await findExistingComposition(draftId, token, [
+        'VALIDATED',
+        'DECLARED',
+        'WAITING_VALIDATION'
```
Yea, exactly. One place that got me confused was [this](https://github.com/opencrvs/opencrvs-core/blob/ocrvs-6208/packages/workflow/src/records/handler/create.ts#L150). From my understanding, `IN_PROGRESS` can never happen as the record cannot be incomplete. But it still leaves us with at least `DECLARED`. `VALIDATED` I can't remember where I got from anymore. 

What do you think? Should we keep `DECLARED`? It might be that the `await createRecord` call below goes through but then one of the proceeding calls fail

rikukissa committed:
fix wrong payload sent to metrics from some workflow handlers

rikukissa committed:
add missing external validation submission to record register handler

rikukissa committed:
add draft id to death and marriage ES documents

rikukissa committed:
remove completed jobs from Redis queue automatically

rikukissa commented on packages/client/src/declarations/createRecord.ts:
```diff
@@ -0,0 +1,96 @@
+   * Now we can start waiting until the record is found in a workqueue
+   */
+
+  let nthTry = 0
+
+  while (true) {
+    const result = await client.query<FetchRecordStatusQuery>({
+      query: FETCH_RECORD_STATUS,
+      fetchPolicy: 'no-cache',
+      variables: {
+        draftId
+      }
+    })
+    if (result.data.fetchRecordStatus.__typename === 'RecordProcessed') {
+      return {
+        recordId: result.data.fetchRecordStatus.recordId,
+        trackingId: result.data.fetchRecordStatus.trackingId,
+        isPotentiallyDuplicate:
+          result.data.fetchRecordStatus.hasPotentialDuplicates
+      }
+    }
+
+    await new Promise((resolve) => setTimeout(resolve, 1000 + nthTry * 1000))
+    nthTry++
+  }
```
In the best case it would go through instantly (as it does now) but definitely a thing worth testing

rikukissa commented on packages/metrics/src/influxdb/client.ts:
```diff
@@ -175,8 +189,53 @@ export const influx = new Influx.InfluxDB({
+
+    for (const measurement of measurements) {
+      const result = await influx.queryRaw(
+        `SELECT * FROM "${measurement}" WHERE "transactionId" = $transactionId;`,
```
Good question. Not sure if InfluxDB is really meant for this :/

rikukissa commented on packages/search/src/features/registration/deduplicate/service.ts:
```diff
@@ -210,6 +211,25 @@ export const searchForBirthDuplicates = async (
         index: OPENCRVS_INDEX_NAME,
         query: {
           bool: {
+            must_not: [
+              {
+                term: { transactionId: transactionId }
+              },
+              ...(body.compositionId
+                ? [
+                    {
+                      /*
+                       * This makes sure that the record is not flagged as
+                       * a duplicate because there are records that have the record being submitted as
+                       * a potential duplicate. This really is only for idempotency and an issue only when record is submitted multiple times.
+                       */
+                      terms: {
+                        'relatesTo.keyword': [body.compositionId]
+                      }
+                    }
+                  ]
+                : [])
+            ],
```
Precisely so the first term makes sure potentially-previously-written record doesn't get flagged as a duplicate when a record is resubmitted. Second clause is to make sure that if the sequence is like 
(all submissions have very similar or same data)

- Submit A (1) – Fails later in the process
- Submit B (1) - Goes through ok, but got flagged as a duplicate of A
- Submit A (2) - Now goes through but gets flagged as a duplicate of B

This use case needs to be explored a bit more I think.

rikukissa committed:
update graphql schema

rikukissa committed:
Merge branch 'develop' into ocrvs-6208

rikukissa commented:
E2Es passing here
https://github.com/opencrvs/e2e/actions/runs/11026468974

naftis commented on packages/commons/src/message-queue/record.ts:
```diff
@@ -0,0 +1,202 @@
+}
+export function useRecordQueue(redisHost: string) {
+  const queue = new Queue<Payload>('records', {
+    connection: { host: redisHost, port: 6379 }
```
allow configuring the port? it's usually allowed in the constants files

rikukissa replied:
Todos updated

naftis replied:
Updated in the confirm/rejectRegistration PR.

naftis commented on packages/metrics/src/features/registration/pointGenerator.test.ts:
```diff
@@ -119,7 +120,7 @@ describe('Verify point generation', () => {
 
     const point = await generateBirthRegPoint(
-      cloneDeep(testPayload),
+      cloneDeep(testPayload) as any,
```
Maybe we could change these to the fixtures in commons so the type would be correct. Needs updating the tests quite a bit though

rikukissa committed:
Merge branch 'develop' into ocrvs-6208

naftis committed:
Merge branch 'develop' into ocrvs-6208

naftis committed:
feat(message-queue): allow confirming and rejecting record externally (#7674)

* feat: allow confirming and rejecting record externally

* revert: no 'patch' needed after all

it would be the correct REST action for the state changes: chose convention for now

* fix: remove dummy definitions

* refactor: improve typing on graphql gateway supported patient identifiers

* refactor: use defined rejection types

* refactor(tests): use defined rejection types

* fix: add some missing rejection reasons

* fix: linter issues

* fix: named types not being included in the generated graphql query

* fix: REVERT THIS? remove trivy from build

* revert: typing work related to reasons

* refactor: stop using global interception in workflow tests

* chore: allow specifying the port to the queue

* feat: add testcontainers

* chore: add host&port to missing places

* refactor: add port and host to redis call

* chore: clean up testcontainers

* fix: use defined redis port in gateway

* Revert "refactor: stop using global interception in workflow tests"

This reverts commit 201cec797c77198a0e891721e2c46ec78ea1c97a.

* revert: yarn.lock changes

* revert: typing efforts in client - add a todo comment to look into this

* Revert "fix: REVERT THIS? remove trivy from build"

This reverts commit 529d702c1fbc25153d919b33e16fb0a1944bfa84.

---------

Co-authored-by: Riku Rouvila <riku.rouvila@gmail.com>
